api/api/accounts/serializers.py

```diff
@@ -25,27 +26,159 @@ class Meta:
 
 
 class UserSerializer(serializers.ModelSerializer):
-    student = StudentSerializer(many=False, allow_null=True, required=False)
-    faculty = FacultySerializer(many=False, allow_null=True, required=False)
-    staff = StaffSerializer(many=False, allow_null=True, required=False)
+    student = StudentSerializer(many=False, allow_null=True, required=False, read_only=False)
+    faculty = FacultySerializer(many=False, allow_null=True, required=False, read_only=False)
+    staff = StaffSerializer(many=False, allow_null=True, required=False, read_only=False)
 
     class Meta:
         model = User
         fields = ['email', 'first_name', 'last_name', 'gender',
                   'contact_no', 'user_type', 'picture_url', 'student', 'faculty', 'staff']
+        extra_kwargs = {'email': {'required': False}}
 
     def create(self, validated_data):
         student_data = validated_data.pop('student', None)
         faculty_data = validated_data.pop('faculty', None)
         staff_data = validated_data.pop('staff', None)
 
+        # check for email
+        if not validated_data.get('email', None):
+            raise serializers.ValidationError('Email field is required.')
+
         user = User.objects.create(**validated_data)
+        user_type = user.user_type
 
+        # create user type if user_type and data match else throw error
         if student_data:
-            Student.objects.create(user=user, **student_data)
+            if user_type == 'Student':
+                Student.objects.create(user=user, **student_data)
+            else:
+                raise serializers.ValidationError('Provided data doesn\'t match with user_type')
         elif faculty_data:
-            Faculty.objects.create(user=user, **faculty_data)
+            if user_type == 'Faculty':
+                Faculty.objects.create(user=user, **faculty_data)
+            else:
+                raise serializers.ValidationError('Provided data doesn\'t match with user_type')
         elif staff_data:
-            Staff.objects.create(user=user, **staff_data)
-
+            if user_type == 'Staff':
+                Staff.objects.create(user=user, **staff_data)
+            else:
+                raise serializers.ValidationError('Provided data doesn\'t match with user_type')
```

Comment:
Instead of throwing and error why don't we just use the `student`, `staff` or `faculty` fields according to the `user_type`?

I think we should avoid throwing errors unless they are essential and make user's life easier/better.
[Reference](https://youtu.be/bmSAYlu0NcY?t=1693)

Reply:
Hmm, that would better for the user. But error should still thrown when user_type and data doesn't match while updating right?

Reply:
> Instead of throwing and error why don't we just use the `student`, `staff` or `faculty` fields according to the `user_type`?
> Wouldn't it be better if I change user_type according to data provided?

Reply:
What if the user puts `student` and `faculty` fields? How will you choose the `user_type`?

Reply:
Yeah, that case couldn't handled. But if user_type is `student` and data provided is of `faculty`, we would just ignore the request?

Reply:
Maybe a mix all both cases? When multiple types of data is provided take according to user type and when there is one type of data but mismatched user_type change user_type and ignore request in any other cases. Would this add unnecessary complexity? 🤔

Reply:
It's an overkill, we want to do the simplest thing that solves the problem. [KISS](https://en.wikipedia.org/wiki/KISS_principle).